[CI]Wait for operator restart after csv patch

gibizer opened the pull request:
The ansible runner image is built by the CI job and the ansibleee-operator csv is patched with the ENV variable to use the freshly built image. This triggers an operator restart which can take more than couple of seconds (we saw 30 secs in some cases). So the CI job needs to wait for the operator to restart before moving forward otherwise it is possible the resources created later will be reconciled by the old operator and therefore uses the old runner image.

cescgina reviewed (approved):
lgtm

gibizer commented:
looking at the result from the minimal job:
```
{"changed": true, "cmd": "set -euxo pipefail\n\n\noc wait -n openstack-operators --timeout=120s --for=delete $(oc get -n openstack-operators pod -l app.kubernetes.io/name=openstack-ansibleee-operator -o name)\n", "delta": "0:02:00.340729", "end": "2024-04-04 09:54:13.161935", "msg": "non-zero return code", "rc": 1, "start": "2024-04-04 09:52:12.821206", "stderr": "++ oc get -n openstack-operators pod -l app.kubernetes.io/name=openstack-ansibleee-operator -o name\n+ oc wait -n openstack-operators --timeout=120s --for=delete pod/openstack-ansibleee-operator-controller-manager-5b7cb794b4cj75d\nerror: timed out waiting for the condition on pods/openstack-ansibleee-operator-controller-manager-5b7cb794b4cj75d", "stderr_lines": ["++ oc get -n openstack-operators pod -l app.kubernetes.io/name=openstack-ansibleee-operator -o name", "+ oc wait -n openstack-operators --timeout=120s --for=delete pod/openstack-ansibleee-operator-controller-manager-5b7cb794b4cj75d", "error: timed out waiting for the condition on pods/openstack-ansibleee-operator-controller-manager-5b7cb794b4cj75d"], "stdout": "", "stdout_lines": []}
```

gibizer commented:
* csv patched at 12:52:12.534359
* ansible waits for pods/openstack-ansibleee-operator-controller-manager-5b7cb794b4cj75d
* but that was started 12:54:46 
that points to a issue with my oc command. If it runs first after the operator is restarted then it will wait for the wrong pod to be deleted and therefore times out.
I need to wait for a different way/.

bogdando commented on tests/roles/dataplane_adoption/tasks/main.yaml:
```diff
@@ -10,6 +10,13 @@
+  ansible.builtin.shell: |
+    {{ shell_header }}
+    {{ oc_header }}
+    oc wait -n openstack-operators --timeout=120s --for=delete $(oc get -n openstack-operators pod -l app.kubernetes.io/name=openstack-ansibleee-operator -o name)
```
this should work short-term. Long term, CI should only wait for some ready status perhaps, and aeee operator should not yield it unless deployed the image. The same story as minor updates for service operators, we don't want eventual consistent playbooks deployed, but only a final version

gibizer replied:
> this should work short-term. Long term, CI should only wait for some ready status perhaps, 

At the moment I don't see any ready status that I can track meaningfully. Deployment has observedGeneration and it changing early (as we expect), but none of the other status fields incl conditions changes while the new operator pod is rolled out. (or I'm missing something). The CSV has ` phase` status field that transition between Installing and Succeeded, but CSV has no observedGeneration field, so I cannot check if it is succeeded on the latest spec update.

> and aeee operator should not yield it unless deployed the image. 

It is not like that the image is deployed right away. The image is used to define jobs for the dataplane services created after the reconfiguration. So I don't see what yielding can be prevented here.

gibizer commented:
```
clusterserviceversion.operators.coreos.com/openstack-ansibleee-operator.v0.0.1 patched (no change)
```
The code needs to be smarter as there are cases when we patch but no changes happens and therefore no Pod restart will happen.

gibizer commented:
```
TASK [dataplane_adoption : use ansible-runner image built from source or latest if none is defined] ***
changed: [localhost] => {"changed": true, "cmd": "set -euxo pipefail\n\n\noc login -u kubeadmin -p 123456789\n\n# openstack-operator catalog pins the sha256, which might differ from latest with time\noc patch csv -n openstack-operators openstack-ansibleee-operator.v0.0.1  --type='json' -p='[{\n  \"op\":\"replace\", \"path\":\"/spec/install/spec/deployments/0/spec/template/spec/containers/1/env/0\",\n  \"value\": {\"name\": \"RELATED_IMAGE_ANSIBLEEE_IMAGE_URL_DEFAULT\", \"value\": \"quay.io/openstack-k8s-operators/openstack-ansibleee-runner:latest\"}}]'\n", "delta": "0:00:00.604663", "end": "2024-04-05 08:03:26.207032", "msg": "", "rc": 0, "start": "2024-04-05 08:03:25.602369", "stderr": "+ oc login -u kubeadmin -p 123456789\n+ oc patch csv -n openstack-operators openstack-ansibleee-operator.v0.0.1 --type=json '-p=[{\n  \"op\":\"replace\", \"path\":\"/spec/install/spec/deployments/0/spec/template/spec/containers/1/env/0\",\n  \"value\": {\"name\": \"RELATED_IMAGE_ANSIBLEEE_IMAGE_URL_DEFAULT\", \"value\": \"quay.io/openstack-k8s-operators/openstack-ansibleee-runner:latest\"}}]'", "stderr_lines": ["+ oc login -u kubeadmin -p 123456789", "+ oc patch csv -n openstack-operators openstack-ansibleee-operator.v0.0.1 --type=json '-p=[{", "  \"op\":\"replace\", \"path\":\"/spec/install/spec/deployments/0/spec/template/spec/containers/1/env/0\",", "  \"value\": {\"name\": \"RELATED_IMAGE_ANSIBLEEE_IMAGE_URL_DEFAULT\", \"value\": \"quay.io/openstack-k8s-operators/openstack-ansibleee-runner:latest\"}}]'"], "stdout": "WARNING: Using insecure TLS client config. Setting this option is not supported!\n\nLogin successful.\n\nYou have access to 72 projects, the list has been suppressed. You can list all projects with 'oc projects'\n\nUsing project \"openstack\".\nclusterserviceversion.operators.coreos.com/openstack-ansibleee-operator.v0.0.1 patched (no change)", "stdout_lines": ["WARNING: Using insecure TLS client config. Setting this option is not supported!", "", "Login successful.", "", "You have access to 72 projects, the list has been suppressed. You can list all projects with 'oc projects'", "", "Using project \"openstack\".", "clusterserviceversion.operators.coreos.com/openstack-ansibleee-operator.v0.0.1 patched (no change)"]}

TASK [dataplane_adoption : Wait for the ansible-operator to restart with the new ENV] ***
skipping: [localhost] => {"changed": false, "false_condition": "\"no change\" not in ansibleee_csv_patched.stdout", "skip_reason": "Conditional result was False"}
```
Now the task waiting for the controller-manager restart is properly skipped if the patching did not change the csv. 
The other case when we need to wait is being tested in https://github.com/openstack-k8s-operators/data-plane-adoption/pull/374 that is rebased on top of this PR.

gibizer commented:
#374 shows that the new runner image is properly 
```
TASK [dataplane_adoption : use ansible-runner image built from source or latest if none is defined] ***
changed: [localhost] => {"changed": true, "cmd": "set -euxo pipefail\n\n\noc login -u kubeadmin -p 123456789\n\n# openstack-operator catalog pins the sha256, which might differ from latest with time\noc patch csv -n openstack-operators openstack-ansibleee-operator.v0.0.1  --type='json' -p='[{\n  \"op\":\"replace\", \"path\":\"/spec/install/spec/deployments/0/spec/template/spec/containers/1/env/0\",\n  \"value\": {\"name\": \"RELATED_IMAGE_ANSIBLEEE_IMAGE_URL_DEFAULT\", \"value\": \"quay.rdoproject.org/openstack-k8s-operators/openstack-ansibleee-runner:42e525ccaa56ed06f00c1ffb8cbc68a9f003bd88\"}}]'\n", "delta": "0:00:00.601583", "end": "2024-04-05 10:20:04.168190", "msg": "", "rc": 0, "start": "2024-04-05 10:20:03.566607", "stderr": "+ oc login -u kubeadmin -p 123456789\n+ oc patch csv -n openstack-operators openstack-ansibleee-operator.v0.0.1 --type=json '-p=[{\n  \"op\":\"replace\", \"path\":\"/spec/install/spec/deployments/0/spec/template/spec/containers/1/env/0\",\n  \"value\": {\"name\": \"RELATED_IMAGE_ANSIBLEEE_IMAGE_URL_DEFAULT\", \"value\": \"quay.rdoproject.org/openstack-k8s-operators/openstack-ansibleee-runner:42e525ccaa56ed06f00c1ffb8cbc68a9f003bd88\"}}]'", "stderr_lines": ["+ oc login -u kubeadmin -p 123456789", "+ oc patch csv -n openstack-operators openstack-ansibleee-operator.v0.0.1 --type=json '-p=[{", "  \"op\":\"replace\", \"path\":\"/spec/install/spec/deployments/0/spec/template/spec/containers/1/env/0\",", "  \"value\": {\"name\": \"RELATED_IMAGE_ANSIBLEEE_IMAGE_URL_DEFAULT\", \"value\": \"quay.rdoproject.org/openstack-k8s-operators/openstack-ansibleee-runner:42e525ccaa56ed06f00c1ffb8cbc68a9f003bd88\"}}]'"], "stdout": "WARNING: Using insecure TLS client config. Setting this option is not supported!\n\nLogin successful.\n\nYou have access to 72 projects, the list has been suppressed. You can list all projects with 'oc projects'\n\nUsing project \"openstack\".\nclusterserviceversion.operators.coreos.com/openstack-ansibleee-operator.v0.0.1 patched", "stdout_lines": ["WARNING: Using insecure TLS client config. Setting this option is not supported!", "", "Login successful.", "", "You have access to 72 projects, the list has been suppressed. You can list all projects with 'oc projects'", "", "Using project \"openstack\".", "clusterserviceversion.operators.coreos.com/openstack-ansibleee-operator.v0.0.1 patched"]}

TASK [dataplane_adoption : Wait for the ansible-operator to restart with the new ENV] ***
changed: [localhost] => {"changed": true, "cmd": "set -euxo pipefail\n\n\noc wait -n openstack-operators --timeout=120s --for=delete pod/openstack-ansibleee-operator-controller-manager-6587547577pzw8d\n", "delta": "0:00:13.409661", "end": "2024-04-05 10:20:17.832543", "msg": "", "rc": 0, "start": "2024-04-05 10:20:04.422882", "stderr": "+ oc wait -n openstack-operators --timeout=120s --for=delete pod/openstack-ansibleee-operator-controller-manager-6587547577pzw8d", "stderr_lines": ["+ oc wait -n openstack-operators --timeout=120s --for=delete pod/openstack-ansibleee-operator-controller-manager-6587547577pzw8d"], "stdout": "pod/openstack-ansibleee-operator-controller-manager-6587547577pzw8d condition met", "stdout_lines": ["pod/openstack-ansibleee-operator-controller-manager-6587547577pzw8d condition met"]}
```

applied now as the new playbook is found an run.
https://logserver.rdoproject.org/74/374/7a1657d7d2784a9d24bc72ac47fb5b7458336bc4/github-check/data-plane-adoption-osp-17-to-extracted-crc/7af707b/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/pods/pre-adoption-validation-openstack-pre-adoption-openstack-2fv2j/logs/openstackansibleee.log

So I think this fix can land now
